model/parts/reserve.py

```diff
@@ -0,0 +1,54 @@
+"""
+Reserve metric and advanced balance calculation
+"""
+
+from model.types import (
+    CryptoAsset,
+    Fiat,
+    Pair,
+    Stable
+)
+
+
+def p_reserve_statistics(
+    _params,
+    _substep,
+    _state_history,
+    prev_state,
+):
+    """
+    calculates reserve statistics
+    """
+    exchange_rates = {
+        Pair(Fiat.USD, Fiat.USD): 1,
+        Pair(Fiat.USD, Fiat.EUR): prev_state['market_price'][Pair(CryptoAsset.CELO, Fiat.EUR)]
+                                  / prev_state['market_price'][Pair(CryptoAsset.CELO, Fiat.USD)],
+        Pair(Fiat.USD, Fiat.BRL): prev_state['market_price'][Pair(CryptoAsset.CELO, Fiat.BRL)]
+                                  /prev_state['market_price'][Pair(CryptoAsset.CELO, Fiat.USD)]
+                      }
+
+    # reserve balance in Fiat.USD
+    reserve_balance = sum([inventory * prev_state['market_price'].get(Pair(key, Fiat.USD))
+                           for key, inventory in prev_state['reserve_balance'].items()])
+
+    stable_fiat_pairs = [[_params['mento_exchanges_config'][exchange].stable,
+                          _params['mento_exchanges_config'][exchange].reference_fiat]
+                         for exchange in _params['mento_exchanges_config']]
+
+    stable_fiat_pairs = dict(zip([pair[0] for pair in stable_fiat_pairs],
+                                 [pair[1] for pair in stable_fiat_pairs]))
+
+    # outstanding stables in Fiat.USD
+    stables_balance = sum([prev_state['floating_supply'].get(stable)
+                           * prev_state['market_price'].get(Pair(stable,
+                                                                 stable_fiat_pairs.get(stable)))
+                           / exchange_rates.get(Pair(Fiat.USD, stable_fiat_pairs.get(stable)))
+                           for stable in Stable])
```

Comment:
This can probably be simplified without the `stable_fiat_pairs` variable.
```python
stables_balance = sum([
                           prev_state['floating_supply'].get(config.stable)
                           * prev_state['market_price'].get(Pair(config.stable, config.reference_fiat))
                           / exchange_rates.get(Pair(Fiat.USD, config.reference_fiat))
                           for (_, config) in params['mento_exchanges_config']
                           ])
```

Don't see the need for the zips and stuff you're doing before or maybe I'm misunderstanding.
Also if you do `dict([['a', 2], ['b', 3]]) == {'a': 2, 'b': 3}`. The zip you're doing is kinda reversing what you're doing with list comprehension inside of the arguments passed to the zip.

Reply:
thanks, that's much cleaner. I was pretty confused with all the nested data structures. It works with a minor adjustment `.items()` to make the config dict iterable: 
```
 stables_balance = sum([
        prev_state['floating_supply'].get(config.stable)
        * prev_state['market_price'].get(Pair(config.stable, config.reference_fiat))
        / exchange_rates.get(Pair(Fiat.USD, config.reference_fiat))
        for (_, config) in _params['mento_exchanges_config'].items()
    ])
```